auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java

```diff
@@ -547,6 +572,94 @@ public MethodSpec createReadMethod(ClassName className,
+            .addModifiers(PRIVATE, STATIC)
+            .returns(String.class)
+            .addParameter(firstCharacter)
+            .addParameter(srcString)
+            .addParameter(indexOfSubstring)
+            .addStatement("return (indexOfSubstring < srcString.length()) ? firstCharacter + srcString.substring(indexOfSubstring) : String.valueOf(firstCharacter)")
+            .build();
+
+    MethodSpec separateCamelCaseMethod = MethodSpec.methodBuilder("separateCamelCase")
+            .addModifiers(PRIVATE, STATIC)
+            .returns(String.class)
+            .addParameter(name)
+            .addParameter(separator)
+            .addStatement("$T translation = new StringBuilder()", StringBuilder.class)
+            .beginControlFlow("for (int i = 0, length = $N.length(); i < length; i++)", name)
+            .addStatement("char character = $N.charAt(i)", name)
+            .beginControlFlow("if ($T.isUpperCase(character) && translation.length() != 0)", Character.class)
+            .addStatement("translation.append($N)", separator)
+            .endControlFlow()
+            .addStatement("translation.append(character)")
+            .endControlFlow()
+            .addStatement("return translation.toString()")
+            .build();
+
+    MethodSpec upperCaseFirstLetterMethod = MethodSpec.methodBuilder("upperCaseFirstLetter")
+            .addModifiers(PRIVATE, STATIC)
+            .returns(String.class)
+            .addParameter(name)
+            .addStatement("$T fieldNameBuilder = new StringBuilder()", StringBuilder.class)
+            .addStatement("int index = 0")
+            .addStatement("char firstCharacter = $N.charAt(index)", name)
+            .addStatement("int length = $N.length()", name)
+            .beginControlFlow("while (index < length - 1)")
+            .beginControlFlow("if ($T.isLetter(firstCharacter))", Character.class)
+            .addStatement("break")
+            .endControlFlow()
+            .addStatement("fieldNameBuilder.append(firstCharacter)")
+            .addStatement("firstCharacter = $N.charAt(++index)", name)
+            .endControlFlow()
+            .beginControlFlow("if (!$T.isUpperCase(firstCharacter))", Character.class)
+            .addStatement("$T modifiedTarget = modifyString($T.toUpperCase(firstCharacter), $N, ++index)", String.class, Character.class, name)
+            .addStatement("return fieldNameBuilder.append(modifiedTarget).toString()")
+            .nextControlFlow("else")
+            .addStatement("return name")
+            .endControlFlow()
+            .build();
+
+    MethodSpec renameFieldMethod = MethodSpec.methodBuilder("renameField")
+            .addModifiers(PUBLIC, STATIC)
+            .returns(String.class)
+            .addParameter(name)
+            .addParameter(gson)
+            .addException(IOException.class)
+            .beginControlFlow("if ($N.fieldNamingStrategy().equals($T.UPPER_CAMEL_CASE))", gson, fieldNamingPolicy)
```

Comment:
This would be cleaner as a switch case, and reduce the number of calls to `fieldNamingStrategy()`.

Reply:
FieldNamingStrategy is an interface, but this PR only supports FieldNamingPolicy, an enum which implements FieldNamingStrategy. I could use the switch statement after first checking if FieldNamingStrategy is an instance of FieldNamingPolicy, would that be appropriate?

Reply:
Makes sense.  While we can skip reflection in the default case, I wonder if it makes sense to support this with reflection for the case where the user uses a non-default naming policy. The reflection code will never be hit unless the user explicitly chooses to use something that requires reflection, i.e. a custom naming policy based on `Field`, making the perf loss a conscious choice.

```java
FieldNamingStrategy strategy = gson.fieldNamingStrategy();
if (strategy instanceof FieldNamingPolicy) {
  switch ((FieldNamingPolicy) strategy) {
    ...
  }
} else {
  return strategy.translateName(MyClass.class.getField("foo"));
}
```